exp/addons/api/v1alpha3/clusterresourceset_types.go

```diff
@@ -0,0 +1,117 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha3
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	// ClusterResourceSetSecretType is the only accepted type of secret in resources
+	ClusterResourceSetSecretType corev1.SecretType = "clusterResourceSet"
+)
+
+// ANCHOR: ClusterResourceSetSpec
+
+// ClusterResourceSetSpec defines the desired state of ClusterResourceSet
+type ClusterResourceSetSpec struct {
+	// Label selector for Clusters. The Clusters that are
+	// selected by this will be the ones affected by this ClusterResourceSet.
+	// It must match the Cluster labels.
+	ClusterSelector metav1.LabelSelector `json:"clusterSelector"`
+	// Resources is a list of Secrets/ConfigMaps where each contains 1 or more resources to be applied to remote clusters.
+	Resources []*ClusterResourceSetResource `json:"resources,omitempty"`
+	// Mode is the mode to be used during applying resources. Defaults to ClusterResourceSetModeApplyOnce.
+	// +optional
+	Mode string `json:"mode"`
+}
+
+// ANCHOR_END: ClusterResourceSetSpec
+
+// ClusterResourceSetResourceKind is a string representation of a ClusterResourceSet resource kind.
+type ClusterResourceSetResourceKind string
+
+const (
+	SecretClusterResourceSetResourceKind    ClusterResourceSetResourceKind = "Secret"
+	ConfigMapClusterResourceSetResourceKind ClusterResourceSetResourceKind = "ConfigMap"
+)
+
+// ClusterResourceSetResource specifies a resource.
+type ClusterResourceSetResource struct {
+	// Name of the resource that is in the same namespace with ClusterResourceSet object.
+	// +kubebuilder:validation:MinLength=1
+	Name string `json:"name,omitempty"`
+	// Kind of the resource. Supported kinds are: Secrets and ConfigMaps.
+	Kind string `json:"kind,omitempty"`
+}
+
+// ClusterResourceSetMode is a string representation of a ClusterResourceSet Mode.
+type ClusterResourceSetMode string
+
+const (
+	// ClusterResourceSetModeApplyOnce is the default mode a ClusterResourceSet mode is assigned by
+	// ClusterResourceSet controller after being created if not specified by user.
+	ClusterResourceSetModeApplyOnce ClusterResourceSetMode = "ApplyOnce"
+)
+
+// SetTypedMode sets the Mode field to the string representation of ClusterResourceSetMode.
+func (c *ClusterResourceSetSpec) SetTypedMode(p ClusterResourceSetMode) {
+	c.Mode = string(p)
+}
+
+// ANCHOR: ClusterResourceSetStatus
+
+// ClusterResourceSetStatus defines the observed state of ClusterResourceSet
+type ClusterResourceSetStatus struct {
+	// ObservedGeneration reflects the generation of the most recently observed MachineSet.
+	// +optional
+	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
+
+	// LastUpdated identifies when the ClusterResourceSet last updated.
+	// +optional
+	LastUpdated *metav1.Time `json:"lastUpdated,omitempty"`
```

Comment:
Where is this used?

Reply:
Nowhere as of now, it is difficult to tell if the ClusterResourceSet is updated during reconcile. Shall we remove status all togethere?

Reply:
I think so, we can use conditions later. For example we might want to add `Ready` when we've applied all the resources to all clusters, and keep updating the timestamp

Reply:
I don't think ready condition is suitable for this object. What will `Ready` mean here? Instead I added `ResourcesApplied` condition. WDYT?